File: old/source/frontend/parser.cpp
#include <frontend/parser.hpp>
#include <config.hpp>

#include <iostream>

// Error message
void Parser::error(string message)
{
    had_error = true;
    current_input->logger.error(message);
}

// Error at token
void Parser::error(Token &token, string message)
{
    had_error = true;
    current_input->logger.error(token, message);
}

// Advance scanner token
void Parser::advance(void)
{
    for (;;)
    {
        if (scanner.scan() != TokenKind::Error)
            return;

        error(scanner.current, scanner.current.to_string());
    }
}

// Match current token kind?
bool Parser::match(TokenKind expected)
{
    if (scanner.current.kind == expected)
    {
        advance();
        return true;
    }

    return false;
}

// Consume token
void Parser::consume(TokenKind expected, string message)
{
    if (match(expected))
        return;

    error(scanner.current, message);
}

// Parse expression
Node *Parser::parse_expression(void)
{
    if (match(TokenKind::Number))
        return new NodeNumber(std::get<double>(scanner.previous.data));

    error(scanner.current, "Unexpected token.");
    return nullptr;
}

// Parse return statement
Node *Parser::parse_return(void)
{
    NodeReturn *node = new NodeReturn();
    node->start      = scanner.previous;

    if (match(TokenKind::Semicolon))
        node->value = nullptr;
    else
    {
        node->value = parse_expression();
        consume(TokenKind::Semicolon, "Expected ';' after return value.");
    }

    return node;
}

// Parse statement
Node *Parser::parse_statement(void)
{
    advance();

    switch (scanner.previous.kind)
    {
        case TokenKind::Return:
            return parse_return();

        default:
        {
            error(scanner.previous, "Unexpected token.");
            return nullptr;
        }
    }
}

// Parse block of statements
Node *Parser::parse_block(void)
{
    NodeBlock *node  = new NodeBlock();
    Token      start = scanner.previous;

    if (!match(TokenKind::RightBrace))
    {
        while (!match(TokenKind::RightBrace) && !match(TokenKind::End))
            node->statements.add(parse_statement());

        if (scanner.previous.kind == TokenKind::End)
            error(start, "Expected '}' after function statement(s).");
    }

    return node;
}

// Parse type
Node *Parser::parse_type(void)
{
    std::string name = scanner.previous.to_string();

    if (types.find(name.c_str()) != types.end())
    {
        Type &type = types[name.c_str()];
        return new NodePrimitive(type.kind);
    }
    
    error(scanner.previous, "Expected type.");
    return nullptr;
}

// Parse function declaration
void Parser::parse_function_declaration(Node *type)
{
    NodeFunction *node = new NodeFunction();
    node->type         = type;

    consume(TokenKind::Identifier, "Expected function name.");
    node->name = scanner.previous;

    // Parse function parameters
    consume(TokenKind::LeftParenthesis, "Expected '(' after function name.");

    if (!match(TokenKind::RightParenthesis))
    {
        do
        {
            NodeParameter *parameter = new NodeParameter();

            // Parse parameter type
            advance();
            parameter->type = parse_type();

            // Parse parameter name
            consume(TokenKind::Identifier, "Expected function parameter name.");
            parameter->name = scanner.previous;

            // Add parameter
            node->parameters.add(parameter);
        }
        while (match(TokenKind::Comma));

        consume(TokenKind::RightParenthesis, "Expected ')' after function parameters.");
    }

    // Parse function block
    consume(TokenKind::LeftBrace, "Expected '{' after ')'.");
    node->block = parse_block();

    current_input->nodes.add(node);
}

// Parse module name
String Parser::parse_module_name(void)
{
    String name = "";

    for (;;)
    {
        if (match(TokenKind::Identifier))
            name += scanner.previous.to_string();
        else
        {
            scanner.current.data = name;
            error(scanner.current, "Invalid module name.");
        }

        if (!match(TokenKind::Dot))
            break;

        name += '.';
    }

    return name;
}

// Parse export
void Parser::parse_export(void)
{
    String module_name = parse_module_name();
    consume(TokenKind::Semicolon, "Expected ';' after module name.");

    // Set input module name
    current_input->module_name = module_name;
}

// Parse import
void Parser::parse_import(void)
{
    NodeImport *node = new NodeImport();
    node->start      = scanner.current;
    node->name       = parse_module_name();
    node->start.data = node->name;
    
    consume(TokenKind::Semicolon, "Expected ';' after module name.");
    current_input->nodes.add(node);
}

// Parse declaration
void Parser::parse_declaration(void)
{
    advance();

    switch (scanner.previous.kind)
    {
        case TokenKind::Identifier:
        {
            Node *type = parse_type();

            if (type)
                parse_function_declaration(type);

            break;
        }

        case TokenKind::Export:
        {
            parse_export();
            break;
        }

        case TokenKind::Import:
        {
            parse_import();
            break;
        }

        default:
            error(scanner.previous, "Unexpected token.");
    }
}

// Parse tokens into AST nodes
bool Parser::parse(void)
{
    for (Input &input : g_inputs)
    {
        current_input = &input;

        scanner.set(input.source.c_str());
        advance();

        while (!match(TokenKind::End))
            parse_declaration();
    }

    return !had_error;
}

// Start parsing
bool Parser::start(void)
{
    // Initialize the types
    types["void"]   = { PrimitiveKind::Void   };
    types["bool"]   = { PrimitiveKind::Bool   };
    types["string"] = { PrimitiveKind::String };

    types["ubyte"]  = { PrimitiveKind::UByte  };
    types["ushort"] = { PrimitiveKind::UShort };
    types["uint"]   = { PrimitiveKind::UInt   };
    types["ulong"]  = { PrimitiveKind::ULong  };

    types["byte"]   = { PrimitiveKind::Byte  };
    types["short"]  = { PrimitiveKind::Short };
    types["int"]    = { PrimitiveKind::Int   };
    types["long"]   = { PrimitiveKind::Long  };

    types["single"] = { PrimitiveKind::Single };
    types["double"] = { PrimitiveKind::Double };

    // Initialize other parser stuff
    had_error = false;

    // Parse tokens into AST nodes
    if (!parse())
        return false;
    
    return true;
}

File: source/frontend/Parser.cs
using System;
using System.Collections.Generic;

namespace Suteki
{
    class Parser
    {
        private Input        CurrentInput;
        private PropertyKind CurrentProperty;

        private Dictionary<string, Type> Types = new Dictionary<string, Type>
        {
            { "void",   new Type(PrimitiveKind.Void)   },
            { "bool",   new Type(PrimitiveKind.Bool)   },
            { "string", new Type(PrimitiveKind.String) },

            { "ubyte",  new Type(PrimitiveKind.UByte)  },
            { "ushort", new Type(PrimitiveKind.UShort) },
            { "uint",   new Type(PrimitiveKind.UInt)   },
            { "ulong",  new Type(PrimitiveKind.ULong)  },

            { "byte",  new Type(PrimitiveKind.Byte)    },
            { "short", new Type(PrimitiveKind.Short)   },
            { "int",   new Type(PrimitiveKind.Int)     },
            { "long",  new Type(PrimitiveKind.Long)    },

            { "single", new Type(PrimitiveKind.Single) },
            { "double", new Type(PrimitiveKind.Double) },
        };

        // Get scanner
        private Scanner Scanner
        {
            get 
            {
                return CurrentInput.Scanner;
            }
        }

        // Get logger
        private Logger Logger
        {
            get
            {
                return CurrentInput.Logger;
            }
        }

        // Get nodes
        private List<Node> Nodes
        {
            get
            {
                return CurrentInput.Nodes;
            }
        }

        // Get previous token
        private Token Previous
        {
            get
            {
                return Scanner.Previous;
            }
        }

        // Get current token
        private Token Current
        {
            get
            {
                return Scanner.Current;
            }
        }

        // Advance current token
        private void Advance()
        {
            for (;;)
            {
                if (Scanner.Next() != TokenKind.Error)
                    break;

                Logger.Error(Current, Current.Data.ToString());
            }
        }

        // Match current token?
        private bool Match(TokenKind expected)
        {
            if (Current.Kind == expected)
            {
                Advance();
                return true;
            }

            return false;
        }

        // Consume token
        private void Consume(TokenKind expected, string message)
        {
            if (Match(expected))
                return;

            Logger.Error(Current, message);
        }

        // Parse module name
        private string ParseModuleName()
        {
            string result = "";

            for (;;)
            {
                if (Match(TokenKind.Identifier))
                    result += Previous.Data.ToString();
                else
                {
                    if (result != "")
                    {
                        Current.Data = result;
                        Logger.Error(Current, "Invalid module name.");
                    }
                    else
                        Logger.Error(Current, "Expected module name.");
                }

                if (!Match(TokenKind.Dot))
                    break;

                result += '.';
            }

            return result;
        }

        // Parse export
        private void ParseExport()
        {
            // Set input module
            string moduleName = ParseModuleName();

            if (CurrentInput.Module != "")
            {
                Previous.Data = moduleName;
                Logger.Error(Previous, "This file was already exported.");
            }

            CurrentInput.Module = moduleName;

            // Optional semicolon
            Match(TokenKind.Semicolon);
        }

        // Parse import
        private void ParseImport()
        {
            // Make node
            NodeImport node            = new NodeImport();
                       node.Start      = Current;
                       node.ModuleName = ParseModuleName();

            Nodes.Add(node);

            // Optional semicolon
            Match(TokenKind.Semicolon);
        }

        // Parse type
        private Node ParseType()
        {
            string typeName = Previous.Data.ToString();

            if (Types.ContainsKey(typeName))
            {
                NodePrimitive node      = new NodePrimitive();
                              node.Kind = Types[typeName].Kind;

                return node;
            }

            return null;
        }

        // Parse expression
        private Node ParseExpression()
        {
            if (Match(TokenKind.Float))
                return new NodeFloat(Previous);

            if (Match(TokenKind.Integer))
                return new NodeInteger(Previous);

            if (Match(TokenKind.String))
                return new NodeString(Previous);

            if (Match(TokenKind.Bool))
                return new NodeBool(Previous);

            Logger.Error(Current, "Unexpected token.");
            return null;
        }

        // Parse return statement
        private Node ParseReturn()
        {
            // Make node
            NodeReturn node       = new NodeReturn();
                       node.Start = Previous;

            // Parse expression
            if (Match(TokenKind.Semicolon))
                node.Expression = null;
            else
            {
                node.Expression = ParseExpression();

                // Optional semicolon
                Match(TokenKind.Semicolon);
            }

            return node;
        }

        // Parse call
        private Node ParseCall(Token name)
        {
            // Make node
            NodeCall node      = new NodeCall();
                     node.Name = name;

            // Parse function parameters
            if (!Match(TokenKind.RightParenthesis))
            {
                do
                {
                    node.Parameters.Add(ParseExpression());
                }
                while (Match(TokenKind.Comma));

                Consume(TokenKind.RightParenthesis, "Expected ')' after function call parameter(s).");
            }

            // Optional semicolon
            Match(TokenKind.Semicolon);

            return node;
        }

        // Parse identifier statement
        private Node ParseIdentifierStatement()
        {
            Token name = Previous;
            
            if (Match(TokenKind.LeftParenthesis))
                return ParseCall(name);
            
            Logger.Error(Current, "Unexpected token.");
            return null;
        }

        // Parse statement
        private Node ParseStatement()
        {
            Advance();

            switch (Previous.Kind)
            {
                case TokenKind.Return:
                    return ParseReturn();

                case TokenKind.Identifier:
                    return ParseIdentifierStatement();

                default:
                {
                    Logger.Error(Previous, "Unexpected token.");
                    return null;
                }
            }
        }

        // Parse block of statements
        private Node ParseBlock()
        {
            Token start = Current;

            // Make node
            NodeBlock node = new NodeBlock();

            // Parse statements
            if (!Match(TokenKind.RightBrace))
            {
                while (!Match(TokenKind.RightBrace) && !Match(TokenKind.End))
                    node.Statements.Add(ParseStatement());

                if (Previous.Kind == TokenKind.End)
                    Logger.Error(start, "Expected '}' after statement(s).");
            }

            return node;
        }

        // Parse function declaration
        private void ParseFunction(Node type, Token name)
        {
            // Make node
            NodeFunction node          = new NodeFunction();
                         node.Property = CurrentProperty;
                         node.Type     = type;
                         node.Name     = name;

            // Parse function parameters
            if (!Match(TokenKind.RightParenthesis))
            {
                do
                {
                    Advance();
                    Node parameterType = ParseType();

                    if (parameterType != null)
                    {
                        Consume(TokenKind.Identifier, "Expected function name after type.");

                        // Make node
                        NodeParameter parameter      = new NodeParameter();
                                      parameter.Type = parameterType;
                                      parameter.Name = Previous;

                        node.Parameters.Add(parameter);
                    }
                    else
                        Logger.Error(Previous, "Expected type.");
                }
                while (Match(TokenKind.Comma));

                Consume(TokenKind.RightParenthesis, "Expected ')' after function parameter(s).");
            }

            // Parse function block
            if (CurrentProperty != PropertyKind.Extern)
            {
                Consume(TokenKind.LeftBrace, "Expected '{'.");
                node.Block = ParseBlock();
            }
            else
            {
                node.Block = null;

                // Optional semicolon
                Match(TokenKind.Semicolon);
            }

            Nodes.Add(node);

            // Reset property
            CurrentProperty = PropertyKind.None;
        }

        // Parse identifier
        private void ParseIdentifier(Node type)
        {
            Consume(TokenKind.Identifier, "Expected identifier after type.");
            Token name = Previous;

            if (Match(TokenKind.LeftParenthesis))
                ParseFunction(type, name);
            else
                Logger.Error(Current, "Unexpected token.");
        }

        // Parse declaration
        private void ParseDeclaration()
        {
            Advance();

            switch (Previous.Kind)
            {
                case TokenKind.Extern:
                {
                    CurrentProperty = PropertyKind.Extern;
                    break;
                }

                case TokenKind.Export:
                {
                    ParseExport();
                    break;
                }

                case TokenKind.Import:
                {
                    ParseImport();
                    break;
                }

                case TokenKind.Identifier:
                {
                    Node type = ParseType();

                    if (type != null)
                        ParseIdentifier(type);
                    else
                        Logger.Error(Previous, "Unexpected token.");

                    break;
                }

                default:
                    break;
            }
        }

        // Start parsing
        public void Start()
        {
            // Parse all source code into AST nodes
            foreach (Input input in Config.Inputs)
            {
                CurrentInput = input;
                Advance();

                while (!Match(TokenKind.End))
                    ParseDeclaration();
            }

            // Don't analyze the nodes if there's an error
            if (Config.HadError)
                return;

            // Analyze the AST nodes
            foreach (Input input in Config.Inputs)
            {
                foreach (Node node in input.Nodes)
                {
                    if (!input.SymbolsAreRegistered)
                        node.RegisterSymbols(input);
                    
                    node.CheckSymbols(input);
                    node.TypeCheck(input);
                    node.Optimize(input);
                }

                input.SymbolsAreRegistered = true;
            }
        }
    }
}